OpenRA.Game/Primitives/Color.cs

```diff
@@ -0,0 +1,361 @@
+			var p = 2 * l - q;
+
+			float[] trgb = { h + 1 / 3.0f, h, h - 1 / 3.0f };
+			float[] rgb = { 0, 0, 0 };
+
+			for (var k = 0; k < 3; k++)
+			{
+				while (trgb[k] < 0) trgb[k] += 1.0f;
+				while (trgb[k] > 1) trgb[k] -= 1.0f;
+			}
+
+			for (var k = 0; k < 3; k++)
+			{
+				if (trgb[k] < 1 / 6.0f)
+					rgb[k] = p + ((q - p) * 6 * trgb[k]);
+				else if (trgb[k] >= 1 / 6.0f && trgb[k] < 0.5)
+					rgb[k] = q;
+				else if (trgb[k] >= 0.5f && trgb[k] < 2.0f / 3)
+					rgb[k] = p + ((q - p) * 6 * (2.0f / 3 - trgb[k]));
+				else
+					rgb[k] = p;
+			}
+
+			return FromArgb(alpha, (int)(rgb[0] * 255), (int)(rgb[1] * 255), (int)(rgb[2] * 255));
+		}
+
+		public static Color FromAhsl(float h, float s, float l)
+		{
+			return FromAhsl(255, h, s, l);
+		}
+
+		public static Color FromAhsv(int alpha, float h, float s, float v)
+		{
+			var ll = 0.5f * (2 - s) * v;
+			var ss = (ll >= 1 || v <= 0) ? 0 : 0.5f * s * v / (ll <= 0.5f ? ll : 1 - ll);
+			return FromAhsl(alpha, h, ss, ll);
+		}
+
+		public static Color FromAhsv(float h, float s, float v)
+		{
+			return FromAhsv(255, h, s, v);
+		}
+
+		public void ToAhsv(out int a, out float h, out float s, out float v)
+		{
+			var ll = 2 * GetBrightness();
+			var ss = GetSaturation() * ((ll <= 1) ? ll : 2 - ll);
+
+			a = A;
+			h = GetHue() / 360f;
+			s = (2 * ss) / (ll + ss);
+			v = (ll + ss) / 2;
+		}
+
+		Color(long argb)
+		{
+			this.argb = argb;
+		}
+
+		public int ToArgb()
+		{
+			return (int)argb;
+		}
+
+		public static Color FromArgb(int alpha, Color baseColor)
+		{
+			return FromArgb(alpha, baseColor.R, baseColor.G, baseColor.B);
+		}
+
+		public static Color FromArgb(int argb)
+		{
+			return FromArgb((byte)(argb >> 24), (byte)(argb >> 16), (byte)(argb >> 8), (byte)argb);
+		}
+
+		public static Color FromArgb(uint argb)
+		{
+			return FromArgb((byte)(argb >> 24), (byte)(argb >> 16), (byte)(argb >> 8), (byte)argb);
+		}
+
+		public static bool TryParse(string value, out Color color)
+		{
+			color = new Color();
+			value = value.Trim();
+			if (value.Length != 6 && value.Length != 8)
+				return false;
+
+			byte red, green, blue, alpha = 255;
+			if (!byte.TryParse(value.Substring(0, 2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out red)
+			    || !byte.TryParse(value.Substring(2, 2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out green)
+			    || !byte.TryParse(value.Substring(4, 2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out blue))
+				return false;
+
+			if (value.Length == 8
+			    && !byte.TryParse(value.Substring(6, 2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out alpha))
+				return false;
+
+			color = FromArgb(alpha, red, green, blue);
+			return true;
+		}
+
+		public static bool operator ==(Color left, Color right)
+		{
+			return left.argb == right.argb;
+		}
+
+		public static bool operator !=(Color left, Color right)
+		{
+			return !(left == right);
+		}
+
+		public float GetBrightness()
```

Comment:
`GetBrightness`/`GetSaturation`/`GetHue` implementations are based on "Go To Definition" on the mono types, then cleaning up the generated code.  Compare with https://github.com/dotnet/corefx/blob/master/src/System.Drawing.Primitives/src/System/Drawing/Color.cs#L489 for the .NET source.